### PR TITLE
Add assert_grafana_dashboards function

### PR DESCRIPTION
**Pull request by rgildein**
Add function assert_grafana_dashboards to check dashboards defined in grafana-dashboard relation data bag. Also add helper function to get all dashboards defined in charm source code.

fixes: #101

I checked the [API](https://grafana.com/docs/agent/latest/static/api/#static-mode-apis-stable) for grafana-agent, and I did not find any endpoint, with which we could check what dashboards will be used.

### CODE REVIEW

File: src/charmed_kubeflow_chisme/testing/cos_integration.py [hunk 1]
```diff
@@ -335,3 +356,31 @@ async def assert_logging(app: Application) -> None:
         assert (
             "endpoint" in unit_data
         ), f"{APP_LOGGING} unit '{unit_name}' relation data are missing 'endpoint'"
+
+
+async def assert_grafana_dashboards(app: Application, dashboards: Set[str]) -> None:
+    """Check Grafana dashboards in relation data bag.
+
+    This function compare dashboards defined in APP_GRAFANA_DASHBOARD relation data bag and
```
Comment (orfeas-k) [inline src/charmed_kubeflow_chisme/testing/cos_integration.py:364]:
```suggestion
    This function compares the dashboards defined in APP_GRAFANA_DASHBOARD relation data bag and
```

File: src/charmed_kubeflow_chisme/testing/cos_integration.py [hunk 2]
```diff
@@ -335,3 +356,31 @@ async def assert_logging(app: Application) -> None:
         assert (
             "endpoint" in unit_data
         ), f"{APP_LOGGING} unit '{unit_name}' relation data are missing 'endpoint'"
+
+
+async def assert_grafana_dashboards(app: Application, dashboards: Set[str]) -> None:
```
Comment (orfeas-k) [inline src/charmed_kubeflow_chisme/testing/cos_integration.py:361]: Should this have an argument for relation name that will default to APP_GRAFANA_DASHBOARD? I think it makes sense since this is configurable in theory, even if all CKF charms use the same relation name.

File: src/charmed_kubeflow_chisme/testing/cos_integration.py [hunk 3]
```diff
@@ -335,3 +356,31 @@ async def assert_logging(app: Application) -> None:
...
+    provided dashboards. e.g. {"my-dashboard-1.json", "my-dashboard-2.json"}
+
+    Args:
+        app (Application): Juju Applicatition object.
+        dashboards (set[str]): Set of dashboard files.
+    """
+    relation_data = await _get_app_relation_data(app, APP_GRAFANA_DASHBOARD)
+    assert (
+        "dashboards" in relation_data
+    ), f"{APP_GRAFANA_DASHBOARD} relation is missing 'dashboards'"
```
Comment (orfeas-k) [inline src/charmed_kubeflow_chisme/testing/cos_integration.py:374]: I think we could rephrase to `relation data is missing 'dashboards'`. WDYT?

File: src/charmed_kubeflow_chisme/testing/cos_integration.py [hunk 4]
```diff
@@ -201,6 +192,24 @@ def _get_alert_rules(data: str) -> Set[str]:
...
+def _get_templates(data: str) -> Set[Dict[str, dict]]:
+    """Get all templates from relation data, where it's defined as string.
+
+    Example of relations data of grafana-dashboard would be:
+
+    ```Python
+    'dashboards': '{"templates": {"file:jupyter-notebook-controller.json": {"content": ...
+    ```
+    """
+    templates_raw = yaml.safe_load(data).get("templates", {})
```
[the remark this answers is not in the document]
Comment (orfeas-k) [reply src/charmed_kubeflow_chisme/testing/cos_integration.py:204]: Ok I saw how this is used so it makes sense. We could though rephrase this to get_dashboard_templates(). This is more of a suggestion/question.

File: src/charmed_kubeflow_chisme/testing/cos_integration.py [hunk 5]
```diff
@@ -201,6 +192,24 @@ def _get_alert_rules(data: str) -> Set[str]:
...
+    templates = {}
+    for key, value in templates_raw.items():
+        file_name = key.replace("file:", "")  # template key is defined as 'file:<file_name>'
+        templates[file_name] = {"charm": value["charm"], "juju_topology": value["juju_topology"]}
+
+    return templates
```
Comment (orfeas-k) [inline src/charmed_kubeflow_chisme/testing/cos_integration.py:210]: Should the function name reflect that this doesn't just fetch the templates but also modifies before returning them?

Comment (orfeas-k) [reply src/charmed_kubeflow_chisme/testing/cos_integration.py:210]: I think it the doctstring at least should reflect how this works

File: tests/unit/testing/test_cos.py [hunk 6]
```diff
@@ -245,11 +248,50 @@ async def test_check_metrics_endpoint(mock_run_on_unit, metrics_endpoint, exp_cm
         ),
     ],
 )
-def test_get_alert_rules(data, exp_alert_rules):
+def test__get_alert_rules(data, exp_alert_rules):
```
Comment (orfeas-k) [inline tests/unit/testing/test_cos.py:251]: is this intentional?

File: tests/unit/testing/test_cos.py [hunk 7]
```diff
@@ -245,11 +248,50 @@ async def test_check_metrics_endpoint(mock_run_on_unit, metrics_endpoint, exp_cm
...
+            '{"templates": {"file:a.json": {"charm": "a", "juju_topology": {}}, '
+            '"file:b.json": {"charm": "b", "juju_topology": {}}}}',
+            {
+                "a.json": {"charm": "a", "juju_topology": {}},
+                "b.json": {"charm": "b", "juju_topology": {}},
+            },
+        ),
+    ],
+)
+def test__get_templates(data, exp_alert_rules):
```
Comment (orfeas-k) [inline tests/unit/testing/test_cos.py:290]:
```suggestion
def test__get_templates(data, exp_templates):
```

File: tests/unit/testing/test_cos.py [hunk 8]
```diff
@@ -245,11 +248,50 @@ async def test_check_metrics_endpoint(mock_run_on_unit, metrics_endpoint, exp_cm
...
     assert _get_alert_rules(data) == exp_alert_rules
 
 
+"""
+
+"""
+
+
+@pytest.mark.parametrize(
+    "data, exp_alert_rules",
```
Comment (orfeas-k) [inline tests/unit/testing/test_cos.py:262]:
```suggestion
    "data, exp_templates",
```

**Review by orfeas-k — COMMENTED** [review]
Good job, left some comments

File: src/charmed_kubeflow_chisme/testing/cos_integration.py (hunk 2 above)
Comment (rgildein) [reply src/charmed_kubeflow_chisme/testing/cos_integration.py:361]: Maybe it's good idea, you can still do it with changing the constant variable. e.g. the [integration tests](https://github.com/canonical/charmed-kubeflow-chisme/blob/4a1a8ab765ed0007ef7c84754c4feff551de0084/tests/integration/test_cos.py#L15-L17)
I think that it's safer to use the constant instead of argument in function, because this constant is used also during deployment and it should be same. We can open a separate issue for this, since if we go with argument, we should add it to all other function also. WDYT?

File: tests/unit/testing/test_cos.py (hunk 6 above)
Comment (rgildein) [reply tests/unit/testing/test_cos.py:251]: Yes, there are two functions `get_alert_rules`  and `_get_alert_rules`. This test is for `_get_alert_rules` and it was skipped due to another test with name `test_get_alert_rules`, which you can see below.

File: src/charmed_kubeflow_chisme/testing/cos_integration.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rgildein) [reply src/charmed_kubeflow_chisme/testing/cos_integration.py:204]: I agree, that it would be more understandable and the name is still not too long, so :+1:.

File: src/charmed_kubeflow_chisme/testing/cos_integration.py (hunk 2 above)
Comment (orfeas-k) [reply src/charmed_kubeflow_chisme/testing/cos_integration.py:365]: Yes agreed on that last point. Do you think we could implement this as part of this PR (since it won't break existing integrations) instead of opening an issue? If not, I 'm OK with issue too. I jsut think it would not be a that big of a change.

Comment (rgildein) [reply src/charmed_kubeflow_chisme/testing/cos_integration.py:365]: You agreed that it's safer to use constant?
I think we should open an issue and discuss it there, I can change it here, but we should also change it for other functions. e.g. `assert_logging` or `assert_metrics_endpoint`

Comment (rgildein) [reply src/charmed_kubeflow_chisme/testing/cos_integration.py:365]: Opened the issue #111.

**Review by orfeas-k — APPROVED** [review]
